core/src/mindustry/game/MapObjectives.java

```diff
@@ -719,58 +731,92 @@ public ShapeTextMarker(){}
 
         @Override
         public void draw(){
-            if(hidden) return;
+            if(hidden || minimap) return;
 
             //in case some idiot decides to make 9999999 sides and freeze the game
             int sides = Math.min(this.sides, 200);
 
-            Lines.stroke(3f, Pal.gray);
-            Lines.poly(pos.x, pos.y, sides, radius + 1f, rotation);
-            Lines.stroke(1f, color);
-            Lines.poly(pos.x, pos.y, sides, radius + 1f, rotation);
+            float scl = autoscale ? 4f / renderer.getDisplayScale() : 1f;
+
+            Draw.z(drawLayer);
+            Lines.stroke(3f * scl, Pal.gray);
+            Lines.poly(pos.x, pos.y, sides, (radius + 1f) * scl, rotation);
+            Lines.stroke(scl, color);
+            Lines.poly(pos.x, pos.y, sides, (radius + 1f) * scl, rotation);
             Draw.reset();
 
             if(fetchedText == null){
                 fetchedText = fetchText(text);
             }
 
             // font size cannot be 0
-            if(Math.abs(fontSize) < 1e-5) return;
+            if(Mathf.equal(fontSize, 0f)) return;
 
-            WorldLabel.drawAt(fetchedText, pos.x, pos.y + radius + textHeight, Draw.z(), flags, fontSize);
+            WorldLabel.drawAt(fetchedText, pos.x, pos.y + radius * scl + textHeight * scl, drawLayer, flags, fontSize * scl);
+        }
+
+        @Override
+        public void drawMinimap(MinimapRenderer minimap){
```

Comment:
There is a massive amount of duplication here. You're essentially copy-pasting the entire draw method for the minimap - why do the markers even need minimap-specific code?

You should not call `draw` *at all* if `hidden` is true, and you should not call `drawMinimap` / `draw` if the booleans to disable them are set. The marker should not be the one handling it in every single override of drawMinimap/draw.

Reply:
About the minimap mode for other markers - it's useful when you want to make a situation awareness system (if the attack system gets too complicated). Default minimap markers blend with terrain when you, for example, display a red marker over crux base, and even if you just set them to other color - it's still not as plain as some sort of unit icon.

I'll better try to cleanup code duplication by merging actual drawing code into one method while leaving position calculations/other stuff as separate functions, maybe even the superclass methods. Will send to re-review when it will be done.